update-vo-config/update-vo-config

```diff
@@ -0,0 +1,507 @@
+                                         "same than the retrieved " +
+                                         "certificate: ignoring it.\n")
+                        old_x509_cert = None
+        sys.stdout.write("Updating template for VOMS server " +
+                         "%s (%s)\n" % (hostname, server_cert_tpl))
+        template = open(server_cert_tpl, 'w')
+        template.write("structure template %s;\n\n" % server_cert_ns)
+        template.write("'cert' ?= <<EOF;\n")
+        if x509_cert:
+            template.write(crypto.dump_certificate(
+                crypto.FILETYPE_PEM, x509_cert
+            ))
+        else:
+            template.write("\n")
+        template.write("EOF\n\n")
+        if old_x509_cert:
+            sys.stdout.write("Existing certificate ('oldcert') found and " +
+                             "different from VO ID card: 'oldcert' defined\n")
+            template.write("'oldcert' ?= <<EOF;\n")
+            template.write(
+                crypto.dump_certificate(crypto.FILETYPE_PEM, old_x509_cert)
+            )
+            template.write("EOF\n\n")
+        template.close()
+
+
+def write_vo_list(vo_details, basedir, namespaces):
+    """Write template containing a list of all defined VOs
+    """
+    vo_list_ns = namespaces['params_dir'] + os.path.sep + 'allvos'
+    vo_list_template = basedir + os.path.sep + vo_list_ns + ".pan"
+    sys.stdout.write("Writing the list of defined VOs (" + vo_list_template +
+                     ")\n")
+    template = open(vo_list_template, 'w')
+    template.write("unique template %s;\n\n" % vo_list_ns)
+    template.write("variable ALLVOS ?= list(\n")
+    for vo_config in sorted(vo_details, key=itemgetter('name')):
+        template.write("    '%s',\n" % vo_config['name'])
+    template.write(");\n\n")
+    template.close()
+
+
+def write_dn_list(vo_details, base_dir, namespaces):
+    """Write template containing subject / issuer of all valid VOMS server
+    certificates
+    """
+    dn_list_ns = namespaces['cert_dir'] + os.path.sep + "voms_dn_list"
+    dn_list_template = base_dir + os.path.sep + dn_list_ns + ".pan"
+    voms_servers = {}
+    sys.stdout.write("Writing the list of defined VOMSs (" +
+                     dn_list_template + ")\n")
+    template = open(dn_list_template, 'w')
+    template.write("unique template " + dn_list_ns + ";\n\n")
+    template.write("variable VOMS_SERVER_DN ?= list(\n")
+    for vo_config in vo_details:
+        if not vo_config['voms_endpoints']:
+            sys.stderr.write("No VOMS server defined in any VO\n")
+        else:
+            for server in vo_config['voms_endpoints']:
+                if server.hostname not in voms_servers:
+                    voms_servers[server.hostname] = server
+    for hostname in sorted(voms_servers.keys()):
+        template.write(write_cert_info(voms_servers[hostname]))
+    template.write(");\n")
+    template.close()
+
+
+def parse_vo_card(vo_card):
```

Comment:
This function seems to be quite big and have a couple of sub-sections, perhaps they could be broken out into sub-functions?

Reply:
The function has been broken into sub-functions.